ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
+ *                         University Research and Technology
+ *                         Corporation.  All rights reserved.
+ * Copyright (c) 2004-2005 The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
+ *                         University of Stuttgart.  All rights reserved.
+ * Copyright (c) 2004-2005 The Regents of the University of California.
+ *                         All rights reserved.
+ * Copyright (c) 2011-2012 Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2015-2019 Research Organization for Information Science
+ *                         and Technology (RIST).  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+#include "ompi/mpi/fortran/mpif-h/bindings.h"
+#include "ompi/mpi/fortran/use-mpi-f08/cdesc/bindings.h"
+#include "ompi/mpi/fortran/base/constants.h"
+
+
+void ompi_send_cdesc(CFI_cdesc_t* x, MPI_Fint *count, MPI_Fint *datatype,
+                     MPI_Fint *dest, MPI_Fint *tag, MPI_Fint *comm, MPI_Fint *ierr)
+{
+    int c_ierr;
+
+    MPI_Comm c_comm = PMPI_Comm_f2c(*comm);
+    MPI_Datatype c_datatype, c_type = PMPI_Type_f2c(*datatype);
+    void *buf = x->base_addr;
+    int c_count =  OMPI_FINT_2_INT(*count);
+
+    c_datatype = c_type;
+    if (x->rank != 0 && !CFI_is_contiguous(x)) {
```

Comment:
What is `CFI_is_contiguous`?  I can't find it anywhere in our code base.  Is it a Fortran compiler-provided function?

Reply:
[Quite interesting...](https://software.intel.com/en-us/fortran-compiler-developer-guide-and-reference-cfi-is-contiguous)

Reply:
Just to be triple sure, we might want to also add a `configure` test for `CFI_is_contiguous`...?  (I know it should probably be there if `ISO_Fortran_binding.h` is there, but it wouldn't hurt to check (because we've seen compile bugs before!).

Reply:
`CFI_is_contiguous()` and the `CFI_cdesc_t` type are defined in `ISO_Fortran_binding.h`. I will have to enhance the test and also run a test program. I do not think the standard specify an ABI, so we should detect some mismatch such as GNU C and Intel Fortran compilers.

Reply:
I don't know if you need to run a test program -- if we can `AC_LINK_IFELSE` (or even just `AC_CHECK_FUNC`), that should probably be enough, right?

Reply:
I do not think the `CFI_cdesc_t` structure and other subroutine are ABI compatibles between compilers, so I believe some more tests are needed. GNU & Fortran compilers do not interoperate (Intel compilers `#define CFI_is_contiguous(dv) for_CFI_is_contiguous(dv)`) so linking C and Fortran binary is enough to detect the incompatibility). An other combination of compilers might link but crash at runtime, so I am afraid that at the end, we have to build __and__ run a test program.

Reply:
Hmm.  I hesitate to `AC_RUN_IFELSE` because of the difficulties that raises with cross-compilation.  If we can avoid `AC_RUN_IFELSE`, we should.